Clean up console code

SteveL-MSFT opened the pull request:
These are features we haven't been supporting already (for PowerShell Core) and no plans to support as they exist for legacy reasons only for Windows PowerShell:

- removed sta/mta switches and code
- removed psconsolefile switch and code
- removed importsystemmodules switch and code
- removed font changing code
- merged `#if CORECLR` code

Fix https://github.com/PowerShell/PowerShell/issues/4987
Fix https://github.com/PowerShell/PowerShell/issues/4350

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleShell.cs:
```diff
@@ -25,18 +24,6 @@ public static int Start(string bannerText, string helpText, string[] args)
         {
             return Start(null, bannerText, helpText, null, args);
```
Do we use first parameter? Can we remove it?

SteveL-MSFT replied:
The first param is RunspaceConfiguration.  It's being removed as part of https://github.com/PowerShell/PowerShell/pull/4942

iSazonov replied:
Thanks! 

Closed.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2937,23 +2893,7 @@ private class ConsoleHostStartupException : Exception
-            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
-            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
-            // reading from Console.In.
-            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
It seems like we should also remove NativeMethods.GetStdHandle() declaration since it is only used here.

SteveL-MSFT replied:
Will remove

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -145,11 +145,6 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
 
         private IntPtr GetMainWindowHandle()
```
Since we now always prompt using console we can get rid of this method since it will never be used.

PaulHigin replied:
PromptForCredential() can be simplified since it now always prompts from console.  Also we can remove the HostUtilities.CredUIPromptForCredential() method since it also will no longer be used.

SteveL-MSFT replied:
Agree, will remove

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -163,42 +158,13 @@ private IntPtr GetMainWindowHandle()
 
         // Determines whether we should prompt using the Console prompting
         // APIs
         private bool PromptUsingConsole()
```
We can remove this method and simply the conditional code in the two places it is called.

SteveL-MSFT replied:
Will change

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2830,20 +2727,7 @@ internal static void SetConsoleTextAttribute(ConsoleHandle consoleHandle, WORD a
-#else
-            CHARSETINFO csi;
-            const DWORD TCI_SRCCODEPAGE = 2;
-            if (!NativeMethods.TranslateCharsetInfo((IntPtr)codePage, out csi, TCI_SRCCODEPAGE))
```
We should remove all unused NativeMethods declarations (TranslateCharsetInfo, GetConsoleWindow, etc.).

SteveL-MSFT replied:
Will remove

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2830,20 +2727,7 @@ internal static void SetConsoleTextAttribute(ConsoleHandle consoleHandle, WORD a
         private static uint CodePageToCharSet(uint codePage)
         {
             const uint OEM_CHARSET = 255;
```
Makes sense, will change

PaulHigin reviewed (approved):
LGTM

iSazonov commented:
@SteveL-MSFT Is the PR ready to merge?

SteveL-MSFT commented:
@iSazonov since this is a breaking change (from Windows PowerShell), let me get an ack from @PowerShell/powershell-committee

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -1893,22 +1893,7 @@ private static bool ReadConsoleOutputCJKSmall
                     else
                     {
-                        int charLength = LengthInBufferCells(contents[r, c].Character);
-                        if (charLength == 2)
```
As I understand this code - somebody is screen scraping, there is a wide character (2 cells), and the font in use can't render the character, so conhost used 1 cell.
This seems like an extreme corner case, but the code seems correct to me and I don't any value in removing the code.
Also - the PR description should be improved - there is no link to an issue providing more detail.

SteveL-MSFT replied:
I'll update the PR description as I did not originally intend for this PR to be so extensive.  I simply started down the path of removing the -mta/-sta code since we weren't using it and then noticed the ConsoleFile code which wasn't being used, then the Font code... ugh...

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -2944,103 +2783,9 @@ private static int LengthInBufferCellsFE(char c, ref HWND hwnd, ref HDC hDC, ref
-        [SuppressMessage("Microsoft.Usage", "CA1806:DoNotIgnoreMethodResults",
-            MessageId = "Microsoft.PowerShell.ConsoleControl+NativeMethods.ReleaseDC(System.IntPtr,System.IntPtr)")]
-        private static int LengthInBufferCells(char c, uint codePage)
-        {
```
I think we still need LengthInBufferCells - but if you want to make a simplifying assumption when you can't know the font, you can use this [code](https://github.com/lzybkr/PSReadLine/blob/8edcf64663d7b7de3d9ef23435473e6529954b1e/PSReadLine/Render.cs#L512) from PSReadline.

SteveL-MSFT replied:
I'll just use the code from PSReadline.  I ended up removing all this code as the fundamental assumption in the code is that we are using a specific codePage so all the calculations ended up returning a width of 1 anyways as all the other supporting code only applied to Windows PowerShell.

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -1682,32 +1657,19 @@ private void DoCreateRunspace(string initialCommand, bool skipProfiles, bool sta
-                runspace.ApartmentState = ApartmentState.STA;
-#endif
-                runspace.ThreadOptions = PSThreadOptions.ReuseThread;
-            }
```
Is this removal intentional? I think if that's not set, we create a new thread every time you execute a new command line.

SteveL-MSFT replied:
Will put back the ReuseThread line

SteveL-MSFT replied:
Actually, this line was only within the `staMode` which was not enabled for the user to set.  So currently, we use the default which is to use a new thread for each pipeline.  Do we want to change this behavior?

lzybkr replied:
`staMode = true` was the default even if the user couldn't set it, right?
At any rate, I do think we want to reuse the thread.

SteveL-MSFT replied:
No, we defaulted to MTA in PSCore

lzybkr replied:
I see. We defaulted to STA on Windows for many years, so reusing the thread seems preferable.

lzybkr replied:
@SteveL-MSFT - are you putting it back?

SteveL-MSFT replied:
I need to verify something on Windows (was using a Mac for the changes) before I put it back.  With my other uber PR to remove apartmentstate all-up, the WSMan provider didn't work correctly.

SteveL-MSFT replied:
Looks good on my Windows box, will run through the tests and push the change.

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2937,23 +2893,7 @@ private class ConsoleHostStartupException : Exception
-            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
-            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
-            // reading from Console.In.
-            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
You should verify the hang doesn't repro in Core. It looks like I did not add a test (probably because the simplest repro required installing Ruby on Windows), but with all the changes in the CLR, it seems prudent to verify we still don't need this code before removing.

---
In reply to: [143062595](https://github.com/PowerShell/PowerShell/pull/4995#discussion_r143062595) [](ancestors = 143062595)

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs:
```diff
@@ -1405,7 +1405,7 @@ int LengthInBufferCells(string s, int offset)
         {
-            return ConsoleControl.LengthInBufferCells(c);
+            return 1;
         }
```
This is wrong, plain and simple.

SteveL-MSFT replied:
WIll fix!

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs:
```diff
@@ -322,24 +322,12 @@ public override
                 // mind is lost. This can lead to a malicious
                 // server prompting the user for a request
                 // which can appear to come from locally.
-                if (!PromptUsingConsole() && desc.ModifiedByRemotingProtocol)
```
Deleting this code makes the comment above very confusing.
It also implies you are introducing a potential security risk.

SteveL-MSFT replied:
Will remove the comment since it only applied to the GUI prompt.

lzybkr reviewed (changes requested):
There are several distinct changes in this PR, some of which don't feel like "clean up".

I think the commits should be cleaned up with better messages (reset and `git add -p`) and when merging, we should not squash this PR.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2937,23 +2893,7 @@ private class ConsoleHostStartupException : Exception
-            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
-            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
-            // reading from Console.In.
-            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
This code wasn't being used since it's in the `#else` statement.

SteveL-MSFT commented:
@lzybkr cleaned up the commits to be distinct groups of work

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -1515,10 +1514,6 @@ private uint Run(string bannerText, string helpText, bool isPrestartWarned, stri
                 _runspaceRef = null;
-                if (staMode) // don't recycle the Runspace in STA mode
-                {
-                    ShouldEndSession = true;
```
If I understand correctly - we default to STA and remove the option to use MTA, correct?

Assuming so, this line should probably stay, but I don't quite understand why it's here.

SteveL-MSFT replied:
No, we defaulted to MTA and removing option to use STA

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:
```diff
@@ -3321,6 +3050,39 @@ internal static void SetConsoleCursorInfo(ConsoleHandle consoleHandle, CONSOLE_C
+
+        private static int LengthInBufferCells(char c)
+        {
+            if (c < 256)
```
This condition only applies to PSReadline.

SteveL-MSFT replied:
will remove

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2937,23 +2893,7 @@ private class ConsoleHostStartupException : Exception
-            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
-            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
-            // reading from Console.In.
-            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
@SteveL-MSFT - I know. That doesn't change my feedback. The code was originally added for Full CLR, but with the significant work that went into Core CLR 2.0, I think we need to revisit the issue in case the code is needed in Core now.

If you remove this code, you'll lose the context, and if it's still an issue, investigation will be much more difficult because nobody will remember the change I made that introduced the problem in the first place.

SteveL-MSFT commented:
@lzybkr I believe your feedback has been addressed, can you take another look?  Thanks

SteveL-MSFT commented:
@PowerShell/powershell-committee discussed this and was fine with all the changes except for removing the STA code.  Resolution is to leave the STA code commented out in case we have valid use cases on Windows on the future requiring support.

SteveL-MSFT committed:
wrapped STA code sections in `#if STAMODE` for potential future use
removed -importsystemmodules switch from powershell.exe and related code
removed -consolefile parameter from powershell.exe and related code

SteveL-MSFT committed:
removed font and codepage handling code that is only applicable to Windows PowerShell
replaced LengthInBufferCells() method with Unicode adapted code from PSReadline

SteveL-MSFT committed:
removed `#if CORECLR` statements merging code base

SteveL-MSFT committed:
[feature]
removed code to show a GUI prompt for credentials as PSCore6 prompts in console

SteveL-MSFT commented:
I believe all feedback has been addressed.  @lzybkr can you take another look when you get a chance?  Thanks.

lzybkr commented:
>         {

This code is not called anymore, and is a slight variant of the version you added from PSReadLine below (which my version should be fine, it's more recent and has better comments.) :)


---
Refers to: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:2736 in 6ca5d51. [](commit_id = 6ca5d517829587785df88ca7c9326adef577c11d, deletion_comment = False)

SteveL-MSFT committed:
Remove unncessary method

SteveL-MSFT commented:
@lzybkr removed that method

iSazonov commented:
@SteveL-MSFT Is the PR ready to merge?

SteveL-MSFT commented:
@iSazonov yes, please!

iSazonov commented:
It seems here we should preserve commits by create a merge commit?

SteveL-MSFT commented:
@iSazonov yes, I believe that is the case
